IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,220 @@
+    for (NSString *key in keys)
+    {
+        if (![self storedCredentialForKey:key])
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
+            if (credential)
+            {
+                [self storeCredential:credential forKey:key];
+            }
+        }
+    }
+}
+
+- (MSIDCredentialCacheItem *)storedCredentialForKey:(NSString *)key
+{
+    __block MSIDCredentialCacheItem *credential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        credential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return credential;
+}
+
+- (void)removeStoredCredentialForKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    if (key.account && key.service)
+    {
+        NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
+        MSIDCredentialCacheItem *credential = [self storedCredentialForKey:credentialKey];
+        if (credential)
+        {
+            return @[credential];
+        }
+        
+        return nil;
+    }
+    
+    NSMutableArray *subPredicates = [[NSMutableArray alloc] init];
+    
+    /*
+     Refresh token either has family Id as 1 or nil.
+     For app refresh token, we pass family id as nil which matches both FRT and RT instead of matching just RT
+     For clients that are not part of family, family is saved as empty string and added to refresh token query.
+     */
+    if ([NSString msidIsStringNilOrBlank:key.familyId])
+    {
+        key.familyId = nil;
+    }
+
+    [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.familyId == %@", key.familyId]];
+    if (key.clientId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.clientId == %@", key.clientId]];
+    if (key.environment)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.environment == %@", key.environment]];
+    if (key.homeAccountId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.homeAccountId == %@", key.homeAccountId]];
+    if (key.credentialType)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.credentialType == %d", key.credentialType]];
+    if (key.realm)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.realm == %@", key.realm]];
+    /*
+     key.target is not added as target can be subset , intersect , superset or exact string match and is matched later in the code.
+     */
+    
+    // Combine all sub-predicates with AND:
+    NSPredicate *matchAttributes = [NSCompoundPredicate andPredicateWithSubpredicates:subPredicates];
+    
+    NSArray *storedCredentials = [self allValues];
+    NSArray *filteredCredentials = [storedCredentials filteredArrayUsingPredicate:matchAttributes];
```

Comment:
just out of curiosity - would it be slower to enumerate dictionary directly here :)? (e.g. enumerateKeysAndObjectsUsingBlock and compare)

Reply:
I think column based search using predicate is faster.

Reply:
Probably, just wondering if we considered all options.